File: packages/aretro/src/aretro/exec/evidence_payload.py
"""Sanitized payload detail document construction for aretro evidence."""

from __future__ import annotations

import hashlib
from pathlib import Path
from typing import Any, Final, Literal, TypeAlias

from asdl_core.clinkr.models import ClinkrModel
from asdl_core.sessions.types import (
    ParsedSession,
    SessionCommandExecution,
    SessionModelEvent,
    SessionSourceRef,
    SessionToolCall,
    SessionToolResult,
    SessionUsage,
    SessionWarning,
)

JsonObject: TypeAlias = dict[str, Any]
CommandMetadata: TypeAlias = dict[str, str | int | bool | None]
_SourceRefKey: TypeAlias = tuple[str | None, str | None, int | None]

MAX_COMMAND_SUBJECT_LENGTH: Final = 500
_COMMAND_SUBJECT_PREFIX_LENGTH: Final = 120
_COMMAND_HASH_PREFIX_LENGTH: Final = 16


class PayloadSourceRefDto(ClinkrModel):
    path: str | None
    uri: str | None
    line_number: int | None


class PayloadWarningDto(ClinkrModel):
    code: str
    message: str
    source_ref: PayloadSourceRefDto | None
    harness: str | None
    adapter_name: str | None


class ModelEventDetailDto(ClinkrModel):
    provider: str | None
    model: str | None
    source_ref: PayloadSourceRefDto | None


class ToolCallDetailDto(ClinkrModel):
    call_id: str | None
    tool_name: str
    argument_keys: tuple[str, ...]
    source_ref: PayloadSourceRefDto | None
    path: str | None
    command_subject: str | None
    command_metadata: CommandMetadata


class ToolResultDetailDto(ClinkrModel):
    tool_call_id: str | None
    tool_name: str | None
    is_error: bool
    has_error_message: bool
    text_length: int | None
    line_count: int | None
    truncated: bool | None
    source_ref: PayloadSourceRefDto | None


class CommandExecutionDetailDto(ClinkrModel):
    command_subject: str
    command_metadata: CommandMetadata
    exit_code: int | None
    cancelled: bool | None
    truncated: bool | None
    output_length: int | None
    line_count: int | None
    source_ref: PayloadSourceRefDto | None


class UsageDetailDto(ClinkrModel):
    input_tokens: int | None
    output_tokens: int | None
    cache_read_tokens: int | None
    cache_write_tokens: int | None
    total_tokens: int | None
    source_ref: PayloadSourceRefDto | None


class SessionDetailDto(ClinkrModel):
    session_index: int
    session_id: str | None
    summary: JsonObject
    model_events: tuple[ModelEventDetailDto, ...]
    tool_calls: tuple[ToolCallDetailDto, ...]
    tool_results: tuple[ToolResultDetailDto, ...]
    command_executions: tuple[CommandExecutionDetailDto, ...]
    usage_events: tuple[UsageDetailDto, ...]
    warnings: tuple[PayloadWarningDto, ...]


class EvidenceDetailItemDto(ClinkrModel):
    evidence_index: int
    item: JsonObject
    supporting_event_pointers: tuple[str, ...]


class AretroEvidencePayloadData(ClinkrModel):
    schema_version: Literal[1] = 1
    repo: JsonObject
    query: JsonObject
    source: JsonObject
    aggregate_metrics: JsonObject
    sessions: tuple[SessionDetailDto, ...]
    warnings: tuple[PayloadWarningDto, ...]
    evidence_items: tuple[EvidenceDetailItemDto, ...]


def build_evidence_payload_data(
    *,
    compact_result: ClinkrModel,
    sessions: tuple[ParsedSession, ...],
) -> AretroEvidencePayloadData:
    """Build a versioned, sanitized detail payload from compact evidence output."""

    compact_data = compact_result.model_dump(mode="json")
    pointer_index: dict[_SourceRefKey, list[str]] = {}
    detail_sessions = tuple(
        _session_detail(
            session=session,
            session_index=session_index,
            compact_session=_json_object_at(compact_data, "sessions", session_index),
            pointer_index=pointer_index,
        )
        for session_index, session in enumerate(sessions)
    )
    evidence_items = tuple(
        _evidence_detail_item(
            evidence_index=evidence_index,
            compact_item=compact_item,
            pointer_index=pointer_index,
        )
        for evidence_index, compact_item in enumerate(_json_objects(compact_data, "evidence_items"))
    )

    return AretroEvidencePayloadData(
        repo=_json_object(compact_data, "repo"),
        query=_json_object(compact_data, "query"),
        source=_json_object(compact_data, "source"),
        aggregate_metrics=_json_object(compact_data, "aggregate_metrics"),
        sessions=detail_sessions,
        warnings=_warnings_from_json(compact_data),
        evidence_items=evidence_items,
    )


def command_subject_for_payload(command: str) -> tuple[str, CommandMetadata]:
    """Return a bounded command subject plus metadata for payload detail."""

    if len(command) <= MAX_COMMAND_SUBJECT_LENGTH:
        return command, {}

    sha256_prefix = hashlib.sha256(command.encode("utf-8")).hexdigest()[
        :_COMMAND_HASH_PREFIX_LENGTH
    ]
    subject = f"{command[:_COMMAND_SUBJECT_PREFIX_LENGTH]}…[sha256:{sha256_prefix}]"
    return subject, {
        "truncated": True,
        "original_length": len(command),
        "sha256_prefix": sha256_prefix,
    }


def _session_detail(
    *,
    session: ParsedSession,
    session_index: int,
    compact_session: JsonObject,
    pointer_index: dict[_SourceRefKey, list[str]],
) -> SessionDetailDto:
    model_events = tuple(
        _model_event_detail(
            event=event,
            pointer=f"/data/sessions/{session_index}/model_events/{event_index}",
            pointer_index=pointer_index,
        )
        for event_index, event in enumerate(session.model_events)
    )
    tool_calls = tuple(
        _tool_call_detail(
            call=call,
            pointer=f"/data/sessions/{session_index}/tool_calls/{call_index}",
            pointer_index=pointer_index,
        )
        for call_index, call in enumerate(session.tool_calls)
    )
    tool_results = tuple(
        _tool_result_detail(
            result=result,
            pointer=f"/data/sessions/{session_index}/tool_results/{result_index}",
            pointer_index=pointer_index,
        )
        for result_index, result in enumerate(session.tool_results)
    )
    command_executions = tuple(
        _command_execution_detail(
            execution=execution,
            pointer=f"/data/sessions/{session_index}/command_executions/{execution_index}",
            pointer_index=pointer_index,
        )
        for execution_index, execution in enumerate(session.command_executions)
    )
    usage_events = tuple(
        _usage_detail(
            usage=usage,
            pointer=f"/data/sessions/{session_index}/usage_events/{usage_index}",
            pointer_index=pointer_index,
        )
        for usage_index, usage in enumerate(session.usage_events)
    )
    warnings = tuple(
        _session_warning_detail(
            warning=warning,
            pointer=f"/data/sessions/{session_index}/warnings/{warning_index}",
            pointer_index=pointer_index,
        )
        for warning_index, warning in enumerate(session.warnings)
    )
    return SessionDetailDto(
        session_index=session_index,
        session_id=session.session_id,
        summary=compact_session,
        model_events=model_events,
        tool_calls=tool_calls,
        tool_results=tool_results,
        command_executions=command_executions,
        usage_events=usage_events,
        warnings=warnings,
    )


def _model_event_detail(
    *,
    event: SessionModelEvent,
    pointer: str,
    pointer_index: dict[_SourceRefKey, list[str]],
) -> ModelEventDetailDto:
    _index_source_ref(event.source_ref, pointer=pointer, pointer_index=pointer_index)
    return ModelEventDetailDto(
        provider=event.provider,
        model=event.model,
        source_ref=_optional_source_ref_to_dto(event.source_ref),
    )


def _tool_call_detail(
    *,
    call: SessionToolCall,
    pointer: str,
    pointer_index: dict[_SourceRefKey, list[str]],
) -> ToolCallDetailDto:
    _index_source_ref(call.source_ref, pointer=pointer, pointer_index=pointer_index)
    command_subject: str | None = None
    command_metadata: CommandMetadata = {}
    if call.command is not None:
        command_subject, command_metadata = command_subject_for_payload(call.command)
    return ToolCallDetailDto(
        call_id=call.call_id,
        tool_name=call.tool_name,
        argument_keys=call.argument_keys,
        source_ref=_optional_source_ref_to_dto(call.source_ref),
        path=call.path,
        command_subject=command_subject,
        command_metadata=command_metadata,
    )


def _tool_result_detail(
    *,
    result: SessionToolResult,
    pointer: str,
    pointer_index: dict[_SourceRefKey, list[str]],
) -> ToolResultDetailDto:
    _index_source_ref(result.source_ref, pointer=pointer, pointer_index=pointer_index)
    return ToolResultDetailDto(
        tool_call_id=result.tool_call_id,
        tool_name=result.tool_name,
        is_error=result.is_error,
        has_error_message=result.error_message is not None,
        text_length=result.text_length,
        line_count=result.line_count,
        truncated=result.truncated,
        source_ref=_optional_source_ref_to_dto(result.source_ref),
    )


def _command_execution_detail(
    *,
    execution: SessionCommandExecution,
    pointer: str,
    pointer_index: dict[_SourceRefKey, list[str]],
) -> CommandExecutionDetailDto:
    _index_source_ref(execution.source_ref, pointer=pointer, pointer_index=pointer_index)
    command_subject, command_metadata = command_subject_for_payload(execution.command)
    return CommandExecutionDetailDto(
        command_subject=command_subject,
        command_metadata=command_metadata,
        exit_code=execution.exit_code,
        cancelled=execution.cancelled,
        truncated=execution.truncated,
        output_length=execution.output_length,
        line_count=execution.line_count,
        source_ref=_optional_source_ref_to_dto(execution.source_ref),
    )


def _usage_detail(
    *,
    usage: SessionUsage,
    pointer: str,
    pointer_index: dict[_SourceRefKey, list[str]],
) -> UsageDetailDto:
    _index_source_ref(usage.source_ref, pointer=pointer, pointer_index=pointer_index)
    return UsageDetailDto(
        input_tokens=usage.input_tokens,
        output_tokens=usage.output_tokens,
        cache_read_tokens=usage.cache_read_tokens,
        cache_write_tokens=usage.cache_write_tokens,
        total_tokens=usage.total_tokens,
        source_ref=_optional_source_ref_to_dto(usage.source_ref),
    )


def _session_warning_detail(
    *,
    warning: SessionWarning,
    pointer: str,
    pointer_index: dict[_SourceRefKey, list[str]],
) -> PayloadWarningDto:
    _index_source_ref(warning.source_ref, pointer=pointer, pointer_index=pointer_index)
    return _warning_to_dto(warning)


def _evidence_detail_item(
    *,
    evidence_index: int,
    compact_item: JsonObject,
    pointer_index: dict[_SourceRefKey, list[str]],
) -> EvidenceDetailItemDto:
    supporting_event_pointers: list[str] = []
    seen_pointers: set[str] = set()
    for source_ref in _source_refs_from_item(compact_item):
        for pointer in pointer_index.get(source_ref, ()):
            if pointer not in seen_pointers:
                supporting_event_pointers.append(pointer)
                seen_pointers.add(pointer)
    return EvidenceDetailItemDto(
        evidence_index=evidence_index,
        item=compact_item,
        supporting_event_pointers=tuple(supporting_event_pointers),
    )


def _index_source_ref(
    source_ref: SessionSourceRef | None,
    *,
    pointer: str,
    pointer_index: dict[_SourceRefKey, list[str]],
) -> None:
    if source_ref is None:
        return
    pointer_index.setdefault(_source_ref_key(source_ref), []).append(pointer)


def _source_ref_key(source_ref: SessionSourceRef) -> _SourceRefKey:
    return (_path_to_string(source_ref.path), source_ref.uri, source_ref.line_number)


def _source_ref_json_key(source_ref: JsonObject) -> _SourceRefKey:
    path = source_ref.get("path")
    uri = source_ref.get("uri")
    line_number = source_ref.get("line_number")
    return (
        path if isinstance(path, str) else None,
        uri if isinstance(uri, str) else None,
        line_number if isinstance(line_number, int) and not isinstance(line_number, bool) else None,
    )


def _source_refs_from_item(item: JsonObject) -> tuple[_SourceRefKey, ...]:
    source_refs = item.get("source_refs")
    if not isinstance(source_refs, list):
        return ()
    return tuple(
        _source_ref_json_key(source_ref)
        for source_ref in source_refs
        if isinstance(source_ref, dict)
    )


def _warning_to_dto(warning: SessionWarning) -> PayloadWarningDto:
    return PayloadWarningDto(
        code=warning.code,
        message=warning.message,
        source_ref=_optional_source_ref_to_dto(warning.source_ref),
        harness=warning.harness,
        adapter_name=warning.adapter_name,
    )


def _warning_from_json(warning: JsonObject) -> PayloadWarningDto:
    source_ref = warning.get("source_ref")
    return PayloadWarningDto(
        code=str(warning.get("code", "")),
        message=str(warning.get("message", "")),
        source_ref=_source_ref_from_json(source_ref) if isinstance(source_ref, dict) else None,
        harness=_optional_string(warning.get("harness")),
        adapter_name=_optional_string(warning.get("adapter_name")),
    )


def _source_ref_to_dto(source_ref: SessionSourceRef) -> PayloadSourceRefDto:
    return PayloadSourceRefDto(
        path=_path_to_string(source_ref.path),
        uri=source_ref.uri,
        line_number=source_ref.line_number,
    )


def _optional_source_ref_to_dto(source_ref: SessionSourceRef | None) -> PayloadSourceRefDto | None:
    if source_ref is None:
        return None
    return _source_ref_to_dto(source_ref)


def _source_ref_from_json(source_ref: JsonObject) -> PayloadSourceRefDto:
    path = source_ref.get("path")
    uri = source_ref.get("uri")
    line_number = source_ref.get("line_number")
    return PayloadSourceRefDto(
        path=path if isinstance(path, str) else None,
        uri=uri if isinstance(uri, str) else None,
        line_number=line_number
        if isinstance(line_number, int) and not isinstance(line_number, bool)
        else None,
    )


def _warnings_from_json(data: JsonObject) -> tuple[PayloadWarningDto, ...]:
    warnings = data.get("warnings")
    if not isinstance(warnings, list):
        return ()
    return tuple(_warning_from_json(warning) for warning in warnings if isinstance(warning, dict))


def _json_object(data: JsonObject, key: str) -> JsonObject:
    value = data.get(key)
    if isinstance(value, dict):
        return value
    return {}


def _json_objects(data: JsonObject, key: str) -> tuple[JsonObject, ...]:
    value = data.get(key)
    if not isinstance(value, list):
        return ()
    return tuple(item for item in value if isinstance(item, dict))


def _json_object_at(data: JsonObject, key: str, index: int) -> JsonObject:
    value = data.get(key)
    if not isinstance(value, list) or index >= len(value):
        return {}
    item = value[index]
    if isinstance(item, dict):
        return item
    return {}


def _optional_string(value: object) -> str | None:
    if isinstance(value, str):
        return value
    return None


def _path_to_string(path: Path | None) -> str | None:
    if path is None:
        return None
    return str(path)

File: ts/packages/aretro/src/payloads/evidence-payload.ts
/**
 * Sanitized payload detail document construction for aretro evidence.
 */

import { z } from "zod";

import { sha256HexPrefix } from "../sha256.ts";

import type {
	AggregateMetricsDto,
	EvidenceItemDto,
	RepoContextDto,
	SessionQueryDto,
	SessionSourceInfoDto,
	SessionSourceRefDto,
	SessionSummaryDto,
	SessionWarningDto,
} from "../contracts.ts";
import type {
	ParsedSession,
	SessionCommandExecution,
	SessionModelEvent,
	SessionSourceRef,
	SessionToolCall,
	SessionToolResult,
	SessionUsage,
	SessionWarning,
} from "../sessions/types.ts";

const MAX_COMMAND_SUBJECT_LENGTH = 500;
const COMMAND_SUBJECT_PREFIX_LENGTH = 120;
const COMMAND_HASH_PREFIX_LENGTH = 16;

type CommandMetadata = Record<string, string | number | boolean | null>;

export const payloadSourceRefDtoSchema = z.object({
	path: z.string().nullable(),
	uri: z.string().nullable(),
	lineNumber: z.number().nullable(),
});

export type PayloadSourceRefDto = z.infer<typeof payloadSourceRefDtoSchema>;

export const payloadWarningDtoSchema = z.object({
	code: z.string(),
	message: z.string(),
	sourceRef: payloadSourceRefDtoSchema.nullable(),
	harness: z.string().nullable(),
	adapterName: z.string().nullable(),
});

export type PayloadWarningDto = z.infer<typeof payloadWarningDtoSchema>;

export const modelEventDetailDtoSchema = z.object({
	provider: z.string().nullable(),
	model: z.string().nullable(),
	sourceRef: payloadSourceRefDtoSchema.nullable(),
});

export type ModelEventDetailDto = z.infer<typeof modelEventDetailDtoSchema>;

export const toolCallDetailDtoSchema = z.object({
	callId: z.string().nullable(),
	toolName: z.string(),
	argumentKeys: z.array(z.string()),
	sourceRef: payloadSourceRefDtoSchema.nullable(),
	path: z.string().nullable(),
	commandSubject: z.string().nullable(),
	commandMetadata: z.record(z.string(), z.union([z.string(), z.number(), z.boolean(), z.null()])),
});

export type ToolCallDetailDto = z.infer<typeof toolCallDetailDtoSchema>;

export const toolResultDetailDtoSchema = z.object({
	toolCallId: z.string().nullable(),
	toolName: z.string().nullable(),
	isError: z.boolean(),
	hasErrorMessage: z.boolean(),
	textLength: z.number().nullable(),
	lineCount: z.number().nullable(),
	isTruncated: z.boolean().nullable(),
	sourceRef: payloadSourceRefDtoSchema.nullable(),
});

export type ToolResultDetailDto = z.infer<typeof toolResultDetailDtoSchema>;

export const commandExecutionDetailDtoSchema = z.object({
	commandSubject: z.string(),
	commandMetadata: z.record(z.string(), z.union([z.string(), z.number(), z.boolean(), z.null()])),
	exitCode: z.number().nullable(),
	isCancelled: z.boolean().nullable(),
	isTruncated: z.boolean().nullable(),
	outputLength: z.number().nullable(),
	lineCount: z.number().nullable(),
	sourceRef: payloadSourceRefDtoSchema.nullable(),
});

export type CommandExecutionDetailDto = z.infer<typeof commandExecutionDetailDtoSchema>;

export const usageDetailDtoSchema = z.object({
	input_tokens: z.number().nullable(),
	output_tokens: z.number().nullable(),
	cache_read_tokens: z.number().nullable(),
	cache_write_tokens: z.number().nullable(),
	total_tokens: z.number().nullable(),
	sourceRef: payloadSourceRefDtoSchema.nullable(),
});

export type UsageDetailDto = z.infer<typeof usageDetailDtoSchema>;

export const sessionDetailDtoSchema = z.object({
	sessionIndex: z.number(),
	sessionId: z.string().nullable(),
	summary: z.record(z.string(), z.unknown()),
	modelEvents: z.array(modelEventDetailDtoSchema),
	toolCalls: z.array(toolCallDetailDtoSchema),
	toolResults: z.array(toolResultDetailDtoSchema),
	commandExecutions: z.array(commandExecutionDetailDtoSchema),
	usageEvents: z.array(usageDetailDtoSchema),
	warnings: z.array(payloadWarningDtoSchema),
});

export type SessionDetailDto = z.infer<typeof sessionDetailDtoSchema>;

export const evidenceDetailItemDtoSchema = z.object({
	evidenceIndex: z.number(),
	item: z.record(z.string(), z.unknown()),
	supportingEventPointers: z.array(z.string()),
});

export type EvidenceDetailItemDto = z.infer<typeof evidenceDetailItemDtoSchema>;

export const aretroEvidencePayloadDataSchema = z.object({
	schemaVersion: z.literal(1),
	repo: z.record(z.string(), z.unknown()),
	query: z.record(z.string(), z.unknown()),
	source: z.record(z.string(), z.unknown()),
	aggregateMetrics: z.record(z.string(), z.unknown()),
	sessions: z.array(sessionDetailDtoSchema),
	warnings: z.array(payloadWarningDtoSchema),
	evidenceItems: z.array(evidenceDetailItemDtoSchema),
});

export type AretroEvidencePayloadData = z.infer<typeof aretroEvidencePayloadDataSchema>;

export interface CompactResult {
	repo: RepoContextDto;
	query: SessionQueryDto;
	source: SessionSourceInfoDto;
	aggregateMetrics: AggregateMetricsDto;
	sessions: readonly SessionSummaryDto[];
	warnings: readonly SessionWarningDto[];
	evidenceItems: readonly EvidenceItemDto[];
}

export function buildEvidencePayloadData(options: {
	compactResult: CompactResult;
	sessions: readonly ParsedSession[];
}): AretroEvidencePayloadData {
	const pointerIndex = new Map<string, string[]>();
	const detailSessions = options.sessions.map((session, sessionIndex) =>
		sessionDetail({
			session,
			sessionIndex,
			compactSession: { ...(options.compactResult.sessions[sessionIndex] ?? {}) },
			pointerIndex,
		}),
	);
	const evidenceItems = options.compactResult.evidenceItems.map((compactItem, evidenceIndex) =>
		evidenceDetailItem({
			evidenceIndex,
			compactItem,
			pointerIndex,
		}),
	);

	return {
		schemaVersion: 1,
		repo: { ...options.compactResult.repo },
		query: { ...options.compactResult.query },
		source: { ...options.compactResult.source },
		aggregateMetrics: { ...options.compactResult.aggregateMetrics },
		sessions: detailSessions,
		warnings: options.compactResult.warnings.map((warning) => payloadWarningFromDto(warning)),
		evidenceItems: evidenceItems,
	};
}

export function commandSubjectForPayload(command: string): {
	subject: string;
	metadata: CommandMetadata;
} {
	if (command.length <= MAX_COMMAND_SUBJECT_LENGTH) {
		return { subject: command, metadata: {} };
	}

	const sha256Prefix = sha256HexPrefix(command, COMMAND_HASH_PREFIX_LENGTH);
	const subject = `${command.slice(0, COMMAND_SUBJECT_PREFIX_LENGTH)}…[sha256:${sha256Prefix}]`;
	return {
		subject,
		metadata: {
			truncated: true,
			originalLength: command.length,
			sha256Prefix: sha256Prefix,
		},
	};
}

function sessionDetail(options: {
	session: ParsedSession;
	sessionIndex: number;
	compactSession: Record<string, unknown>;
	pointerIndex: Map<string, string[]>;
}): SessionDetailDto {
	const { session, sessionIndex, compactSession, pointerIndex } = options;
	const modelEvents = session.model_events.map((event, eventIndex) =>
		modelEventDetail({
			event,
			pointer: `/data/sessions/${sessionIndex}/modelEvents/${eventIndex}`,
			pointerIndex,
		}),
	);
	const toolCalls = session.tool_calls.map((call, callIndex) =>
		toolCallDetail({
			call,
			pointer: `/data/sessions/${sessionIndex}/toolCalls/${callIndex}`,
			pointerIndex,
		}),
	);
	const toolResults = session.tool_results.map((result, resultIndex) =>
		toolResultDetail({
			result,
			pointer: `/data/sessions/${sessionIndex}/toolResults/${resultIndex}`,
			pointerIndex,
		}),
	);
	const commandExecutions = session.command_executions.map((execution, executionIndex) =>
		commandExecutionDetail({
			execution,
			pointer: `/data/sessions/${sessionIndex}/commandExecutions/${executionIndex}`,
			pointerIndex,
		}),
	);
	const usageEvents = session.usage_events.map((usage, usageIndex) =>
		usageDetail({
			usage,
			pointer: `/data/sessions/${sessionIndex}/usageEvents/${usageIndex}`,
			pointerIndex,
		}),
	);
	const warnings = session.warnings.map((warning, warningIndex) =>
		sessionWarningDetail({
			warning,
			pointer: `/data/sessions/${sessionIndex}/warnings/${warningIndex}`,
			pointerIndex,
		}),
	);
	return {
		sessionIndex: sessionIndex,
		sessionId: session.session_id,
		summary: compactSession,
		modelEvents: modelEvents,
		toolCalls: toolCalls,
		toolResults: toolResults,
		commandExecutions: commandExecutions,
		usageEvents: usageEvents,
		warnings,
	};
}

function modelEventDetail(options: {
	event: SessionModelEvent;
	pointer: string;
	pointerIndex: Map<string, string[]>;
}): ModelEventDetailDto {
	indexSourceRef(options.event.source_ref ?? null, {
		pointer: options.pointer,
		pointerIndex: options.pointerIndex,
	});
	return {
		provider: options.event.provider ?? null,
		model: options.event.model ?? null,
		sourceRef: optionalSourceRefToDto(options.event.source_ref ?? null),
	};
}

function toolCallDetail(options: {
	call: SessionToolCall;
	pointer: string;
	pointerIndex: Map<string, string[]>;
}): ToolCallDetailDto {
	indexSourceRef(options.call.source_ref, {
		pointer: options.pointer,
		pointerIndex: options.pointerIndex,
	});
	let commandSubject: string | null = null;
	let commandMetadata: CommandMetadata = {};
	if (options.call.command !== undefined) {
		const bounded = commandSubjectForPayload(options.call.command);
		commandSubject = bounded.subject;
		commandMetadata = bounded.metadata;
	}
	return {
		callId: options.call.call_id,
		toolName: options.call.tool_name,
		argumentKeys: [...options.call.argument_keys],
		sourceRef: optionalSourceRefToDto(options.call.source_ref),
		path: options.call.path ?? null,
		commandSubject: commandSubject,
		commandMetadata: commandMetadata,
	};
}

function toolResultDetail(options: {
	result: SessionToolResult;
	pointer: string;
	pointerIndex: Map<string, string[]>;
}): ToolResultDetailDto {
	indexSourceRef(options.result.source_ref, {
		pointer: options.pointer,
		pointerIndex: options.pointerIndex,
	});
	return {
		toolCallId: options.result.tool_call_id,
		toolName: options.result.tool_name,
		isError: options.result.is_error,
		hasErrorMessage: options.result.error_message !== null,
		textLength: options.result.text_length,
		lineCount: options.result.line_count,
		isTruncated: options.result.truncated,
		sourceRef: optionalSourceRefToDto(options.result.source_ref),
	};
}

function commandExecutionDetail(options: {
	execution: SessionCommandExecution;
	pointer: string;
	pointerIndex: Map<string, string[]>;
}): CommandExecutionDetailDto {
	indexSourceRef(options.execution.source_ref, {
		pointer: options.pointer,
		pointerIndex: options.pointerIndex,
	});
	const bounded = commandSubjectForPayload(options.execution.command);
	return {
		commandSubject: bounded.subject,
		commandMetadata: bounded.metadata,
		exitCode: options.execution.exit_code,
		isCancelled: options.execution.cancelled,
		isTruncated: options.execution.truncated,
		outputLength: options.execution.output_length,
		lineCount: options.execution.line_count,
		sourceRef: optionalSourceRefToDto(options.execution.source_ref),
	};
}

function usageDetail(options: {
	usage: SessionUsage;
	pointer: string;
	pointerIndex: Map<string, string[]>;
}): UsageDetailDto {
	indexSourceRef(options.usage.source_ref, {
		pointer: options.pointer,
		pointerIndex: options.pointerIndex,
	});
	return {
		input_tokens: options.usage.input_tokens,
		output_tokens: options.usage.output_tokens,
		cache_read_tokens: options.usage.cache_read_tokens,
		cache_write_tokens: options.usage.cache_write_tokens,
		total_tokens: options.usage.total_tokens,
		sourceRef: optionalSourceRefToDto(options.usage.source_ref),
	};
}

function sessionWarningDetail(options: {
	warning: SessionWarning;
	pointer: string;
	pointerIndex: Map<string, string[]>;
}): PayloadWarningDto {
	indexSourceRef(options.warning.source_ref, {
		pointer: options.pointer,
		pointerIndex: options.pointerIndex,
	});
	return warningToDto(options.warning);
}

function evidenceDetailItem(options: {
	evidenceIndex: number;
	compactItem: EvidenceItemDto;
	pointerIndex: Map<string, string[]>;
}): EvidenceDetailItemDto {
	const supportingEventPointers: string[] = [];
	const seenPointers = new Set<string>();
	for (const sourceRef of options.compactItem.sourceRefs) {
		const key = refKey(sourceRef.path, sourceRef.uri, sourceRef.lineNumber);
		const pointers = options.pointerIndex.get(key) ?? [];
		for (const pointer of pointers) {
			if (!seenPointers.has(pointer)) {
				supportingEventPointers.push(pointer);
				seenPointers.add(pointer);
			}
		}
	}
	return {
		evidenceIndex: options.evidenceIndex,
		item: evidenceItemToPayloadObject(options.compactItem),
		supportingEventPointers: supportingEventPointers,
	};
}

function indexSourceRef(
	sourceRef: SessionSourceRef | null,
	options: { pointer: string; pointerIndex: Map<string, string[]> },
): void {
	if (sourceRef === null) {
		return;
	}
	const key = refKey(sourceRef.path, sourceRef.uri, sourceRef.line_number);
	const pointers = options.pointerIndex.get(key);
	if (pointers === undefined) {
		options.pointerIndex.set(key, [options.pointer]);
	} else {
		pointers.push(options.pointer);
	}
}

function refKey(path: string | null, uri: string | null, lineNumber: number | null): string {
	return JSON.stringify([path, uri, lineNumber]);
}

function warningToDto(warning: SessionWarning): PayloadWarningDto {
	return {
		code: warning.code,
		message: warning.message,
		sourceRef: optionalSourceRefToDto(warning.source_ref),
		harness: warning.harness,
		adapterName: warning.adapter_name,
	};
}

function payloadWarningFromDto(warning: SessionWarningDto): PayloadWarningDto {
	return {
		code: warning.code,
		message: warning.message,
		sourceRef: warning.sourceRef === null ? null : dtoSourceRefToPayload(warning.sourceRef),
		harness: warning.harness,
		adapterName: warning.adapterName,
	};
}

function evidenceItemToPayloadObject(item: EvidenceItemDto): Record<string, unknown> {
	return {
		kind: item.kind,
		subject: item.subject,
		summary: item.summary,
		count: item.count,
		sessionCount: item.sessionCount,
		sourceRefs: item.sourceRefs.map((sourceRef) => dtoSourceRefToPayload(sourceRef)),
		metadata: { ...item.metadata },
	};
}

function optionalSourceRefToDto(sourceRef: SessionSourceRef | null): PayloadSourceRefDto | null {
	if (sourceRef === null) {
		return null;
	}
	return {
		path: sourceRef.path,
		uri: sourceRef.uri,
		lineNumber: sourceRef.line_number,
	};
}

function dtoSourceRefToPayload(sourceRef: SessionSourceRefDto): PayloadSourceRefDto {
	return {
		path: sourceRef.path,
		uri: sourceRef.uri,
		lineNumber: sourceRef.lineNumber,
	};
}
